third_party/move/move-compiler-v2/src/pipeline/control_flow_graph_simplifier.rs

```diff
@@ -0,0 +1,595 @@
+        debug_assert!(!self
+            .0
+            .successors
+            .get(&block_to_remove)
+            .expect("successors")
+            .contains(&block_to_remove));
+        let from = self.0.get_block_label(block_to_remove).expect("label");
+        let to = self.0.get_block_label(redirect_to).expect("label");
+        self.0.remove_block(
+            block_to_remove,
+            |this, pred| {
+                // for all predecessors of `block_to_remove`, let them jump to `redirect_to` instead
+                if pred != this.entry_block {
+                    let code = this.code_blocks.get_mut(&pred).expect("code block");
+                    Self::redirects_block(code, from, to);
+                }
+                // update successors of pred by replacing `block_to_remove` with `redirect_to`
+                this.replace_succs(pred, block_to_remove, redirect_to);
+            },
+            // replace `block_to_remove` by predecessors of `block_to_remove`
+            // in predecessors of `redirect_to`
+            |this, succ| {
+                debug_assert!(succ == redirect_to);
+                this.remove_pred(succ, block_to_remove);
+                let mut preds_of_block_to_remove = this.preds(block_to_remove).to_vec();
+                this.preds_mut(succ).append(&mut preds_of_block_to_remove);
+            },
+        );
+    }
+
+    /// Redirects a code block so that it jumps/branches to `to`
+    /// where it originally jumps/branches to `from`.
+    /// Does nothing if `code` doesn't end with a jump/branch
+    /// Requries: `code` not empty
+    fn redirects_block(code: &mut [Bytecode], from: Label, to: Label) {
+        let last_instr = code.last_mut().expect("last instruction");
+        match last_instr {
+            Bytecode::Branch(_, l0, l1, _) => {
+                subst_label(l0, from, to);
+                subst_label(l1, from, to);
+            },
+            Bytecode::Jump(_, label) => {
+                subst_label(label, from, to);
+            },
+            _ => {},
+        }
+    }
+}
+
+/// Transformation state for transformation 2 in the module doc
+struct RedundantJumpRemover(pub ControlFlowGraphCodeGenerator);
+
+impl RedundantJumpRemover {
+    /// Wrapper
+    pub fn new(cfg_generator: ControlFlowGraphCodeGenerator) -> Self {
+        Self(cfg_generator)
+    }
+
+    /// Performs the transformation 2 described in the module doc
+    pub fn transform(&mut self) {
+        for block in self.0.blocks() {
+            // the later condition says that `block` is unreachable or has been removed
+            if self.0.is_trivial_block(block) || !self.0.predecessors.contains_key(&block) {
+                continue;
+            }
+            self.transform_edges_from(block)
+        }
+    }
+
+    /// Removes all redundant edges from `block`
+    /// Requires: `block` still in the cfg, `block` is not entry/exit
+    fn transform_edges_from(&mut self, block: BlockId) {
```

Comment:
Maybe rename this function to `remove_redundant_edges` or something more informative about what this function does?

Reply:
done